services/middleware/src/transactions.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Transaction logic for `MiddlewareService`.
+
+use exonum::runtime::{
+    rust::{CallContext, ChildAuthorization},
+    AnyTx, ExecutionError, InstanceStatus,
+};
+use exonum_derive::*;
+use exonum_proto::ProtobufConvert;
+use semver::VersionReq;
+use serde_derive::*;
+
+use crate::{proto, MiddlewareService};
+
+/// Errors of the `MiddlewareService`.
+#[derive(Debug, Clone, Copy, ExecutionFail)]
+pub enum Error {
+    /// The service instance targeted by the checked call does not exist.
+    NoService = 0,
```

Comment:
Interesting side effect: the direct call to the non-existent service will be considered malformed early and won't be stored, but if it was correctly proxied through `middleware`, then the `CheckedCall` transaction will be persisted.

Reply:
Yes, that's a downside of this approach. It could be potentially dealt with by the in-pool transaction verification, but this feature is not in the nearest roadmap.